### DISCUSSION TITLE
Targeting certain users for Push Notifications

### DISCUSSION THREAD

**Discussion by wh1337** [ShafSpecs/remix-pwa#78] [General]
I was just curious, is there a way that we can target only certain users / devices for Push Notifications? I looked around but couldn't find any mention of it. 

What I'm looking to do is only target user `a, b, and c` when something happens on the server, versus sending a push notification to everyone. I may be miss understanding how the Push Notifications API works in `remix-pwa` but just wanted to double check.

Thanks!

**Comment by ShafSpecs** [comment]
You should be able to use Push API to target specific users from the server, but if for example you want the users to listen to the event always (e.g a messaging app), a better alternative in my opinion is to use WebSockets (remix has an example repo for adding it) and when an event is triggered on the server (e.g message sent from A to B), you just use the Notification API to alert the client. This is more versatile imo cause you can customise how the notification is presented to the user

**Reply by ShafSpecs** [reply]
But if for example you wanted to notify a server event (e.g When a user uses over $5, send an email), you could trigger a push on the server to your emailing api and pass extra info to the push payload (list of users that have used over $5) and then, you send an email to those selected users. Push API is to be used server-side anyways

**Reply by wh1337** [reply]
Thank you for the explanation! Does the Notifications API work in the PWA, say on an iOS device? Or does the user have to have the application open? Looking to leverage remix-pwa and the new ability in iOS 16 to have push notifications. Hope that makes sense.

**Reply by ShafSpecs** [reply]
https://developer.mozilla.org/en-US/docs/Web/API/Notification#browser_compatibility

Seems there's a difference between Safari on iOS and Safari. Not sure which is which, but the support is currently shaky on iOS for PWA. I know they are taking steps to improve that though. As for wether the app needs to be open before a notification pops up, it doesn't have to be open.

**Comment by ShafSpecs** [comment]
@wh1337 Thanks to WWDC23, can confirm it does work on iOS! Adding this new guide here, feel free to go through all the docs and get familiar with remix-pwa modules

https://remix-pwa-docs.vercel.app/push/chat-push
